collectors/builtin/spark.py

```
@@ -0,0 +1,373 @@
+# [
+#     {
+#         "id": "driver",
+#         "hostPort": "10.0.2.15:33870",
+#         "rddBlocks": 99,
+#         "memoryUsed": 98,
+#         "diskUsed": 97,
+#         "activeTasks": 96,
+#         "failedTasks": 95,
+#         "completedTasks": 94,
+#         "totalTasks": 93,
+#         "totalDuration": 92,
+#         "totalInputBytes": 91,
+#         "totalShuffleRead": 90,
+#         "totalShuffleWrite": 89,
+#         "maxMemory": 278019440,
+#         "executorLogs": {}
+#     }
+# ]
+EXECUTOR_METRICS = [
+    'rddBlocks',
+    'memoryUsed',
+    'diskUsed',
+    'activeTasks',
+    'failedTasks',
+    'completedTasks',
+    'totalTasks',
+    'totalDuration',
+    'totalInputBytes',
+    'totalShuffleRead',
+    'totalShuffleWrite',
+    'maxMemory'
+]
+
+# spacke rdd metric
+# [
+#     {
+#         "id": 6,
+#         "name": "PythonRDD",
+#         "numPartitions": 2,
+#         "numCachedPartitions": 2,
+#         "storageLevel": "Memory Serialized 1x Replicated",
+#         "memoryUsed": 284,
+#         "diskUsed": 0
+#     }
+# ]
+RDD_METRCIS = [
+    'numPartitions',
+    'numCachedPartitions',
+    'memoryUsed',
+    'diskUsed'
+]
+
+
+class Spark(CollectorBase):
+    def __init__(self, config, logger, readq):
+        super(Spark, self).__init__(config, logger, readq)
+
+        self.spark_cluster_mode = self.get_config('spark_cluster_mode', SPARK_STANDALONE_MODE)
+        self.host = self.get_config('spark_host', "localhost")
+        self.port = self.get_config('spark_port', 8080)
+        self.spark_url = 'http://%s:%s' % (self.host, self.port)
+
+    def __call__(self):
+        spark_apps = self._get_running_apps()
+        self._spark_job_metrics(spark_apps)
+        self._spark_stage_metrics(spark_apps)
+        self._spark_executor_metrics(spark_apps)
+
+    def _get_running_apps(self):
+        if self.spark_cluster_mode is None:
+            self.log_exception('spark_cluster_node must not be null')
+            return None
```

how to handle None, should you fail fast

---

yes, if return none and fail fast .
And  I have some question while I was writing the collector .
After our customer deploy collector with a section script , and we haven't  git any configure interface out of them . Do they need to open config file and overwrite the property?

---

yes, currently we have to manually config. Hopefully default value will serve most cases.